tests/usm/usm_api.h

```diff
@@ -0,0 +1,688 @@
+      sycl::buffer<T, 1> buf(values.data(), range);
+
+      queue.submit([&](sycl::handler &cgh) {
+        auto acc = buf.template get_access<sycl::access_mode::write>(cgh);
+        cgh.parallel_for<kernel_storage_copy<T, count, alloc>>(
+            range, [=](sycl::id<1> idx) {
+              const size_t linearIndex = idx[0];
+              acc[idx] = ptr[linearIndex];
+            });
+      });
+      // buffer destructor waits for the kernel submitted
+    }
+    return values;
+  }
+};
+
+/** @brief Encapsulates the method-specific logic, for example logic specific
+ *         to the memcpy() and memset() tests
+ */
+namespace tests {
+
+/** @brief Encapsulates the implementation details
+ */
+namespace detail {
+
+/** @brief Provides verification for the given pointer data against the given
+ *         reference values
+ */
+template <size_t count, typename T, class ReferenceIt>
+void check_values(const T *begin, const ReferenceIt &reference) {
+  const auto end = begin + count;
+  auto diff = std::mismatch(begin, end, reference);
+  if (diff.first != end) {
+    std::string message{"Mismatch on "};
+    message += std::to_string(end - begin);
+    message += ": retrieved " + std::to_string(*diff.first);
+    message += ", expected " + std::to_string(*diff.second);
+    throw sycl_cts::util::fail_check(message);
+  }
+}
+
+/** @brief Provides generic copy test logic for memcpy() and copy() tests
+ *  @tparam sourceAllocation Allocation type to use for data source
+ */
+template <typename T, size_t count, allocation sourceAllocation>
+class copyGeneric {
+ protected:
+  using storage_t = storage<T, count, sourceAllocation>;
+  const typename storage_t::type source;
+
+  std::array<T, count> reference;
+
+ public:
+  copyGeneric(sycl::queue &queue, int seed)
+      : source(storage_t::get(queue, T{seed})) {
+    auto values = storage_t::copy_values(queue, source);
+    reference.swap(values);
+  }
+
+  void check(const T *ptr) { check_values<count>(ptr, reference.begin()); }
+};
+
+/** @brief Provides generic test logic for the copy() method tests
+ */
+template <typename T, size_t count, allocation sourceAllocation>
+class copy : public copyGeneric<T, count, sourceAllocation> {
+  using baseT = copyGeneric<T, count, sourceAllocation>;
+
+ public:
+  copy(sycl::queue &queue, int seed) : baseT(queue, seed) {}
+
+  /** @brief The copy() method supports non-USM pointers
+   */
+  static constexpr bool has_non_usm_support() { return true; }
+
+  template <allocation alloc>
+  static std::string description() {
+    return "copy from " + get_allocation_decription<sourceAllocation>() +
+           " to " + get_allocation_decription<alloc>();
+  }
+
+  template <typename parentT, typename... depEventsT>
+  auto call(parentT &parent, T *ptr, depEventsT &&... events) const {
+    return parent.copy(this->source.get(), ptr, count, events...);
+  }
+};
+
+/** @brief Provides generic test logic for the memcpy() method tests
```

Comment:
```suggestion
/** @brief Provides generic test logic for the memcpy() member function tests
```

Reply:
I've replaced "method" to "member function" in this commit: 800816219c8f8b2a87887df4ef54c8088b0a153b